pkg/util/startup/startup_test.go

```diff
@@ -0,0 +1,281 @@
+	"github.com/cockroachdb/cockroach/pkg/security/securitytest"
+	"github.com/cockroachdb/cockroach/pkg/server"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMain(m *testing.M) {
+	securityassets.SetLoader(securitytest.EmbeddedAssets)
+	serverutils.InitTestServerFactory(server.TestServerFactory)
+	serverutils.InitTestClusterFactory(testcluster.TestClusterFactory)
+	os.Exit(m.Run())
+}
+
+// TestStartupFailure can help reproducing failures found in randomized test as
+// it may be tricky to get the same results easily. To reproduce failure, use
+// failing range from test and plug in into faultKey.
+// As of time of writing following ranges trigger retries:
+// - /System/NodeLivenessMax (keys.BootstrapVersionKey)
+// - /Table/11 (keys.SystemSQLCodec.TablePrefix(11))
+// If startup.IsRetryableReplicaError is changed to return false every time,
+// test should fail if range from above is used.
+func TestStartupFailure(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	skip.IgnoreLint(t, "this test should be used to reproduce failures from random tests")
+
+	// Set faultyKey to suspected range that is causing startup failure, comment
+	// out skip statement above and run the test.
+	faultyKey := keys.BootstrapVersionKey
+	runCircuitBreakerTestForKey(t,
+		func(spans []roachpb.Span) (good, bad []roachpb.Span) {
+			for _, span := range spans {
+				if span.ContainsKey(faultyKey) {
+					bad = append(bad, span)
+				} else {
+					good = append(good, span)
+				}
+			}
+			return good, bad
+		})
+}
+
+// TestStartupFailureRandomRange verifies that no range with transient failure
+// would cause node startup to fail.
+// To achieve this, test picks a random range then relocates replicas of the
+// range so that restarting a node would cause this range to lose quorum,
+// stops node, ensures circuit breaker is set and starts node back. The node
+// should start without raising error (for detailed explanation see doc on
+// runCircuitBreakerTestForKey).
+// If this test fails, error should indicate which range failed and hopefully
+// error will also contain call stack that caused failure. This range could be
+// used in TestStartupFailure to investigate code path not covered by startup
+// retries.
+func TestStartupFailureRandomRange(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	skip.UnderStressRace(t, "6 nodes with replication is too slow for stress race")
```

Comment:
// This test takes 30s and so we don't want it to run in the "blocking path" of CI at all, and we also don't want to stress it in nightlies as part of a big package (where it will take a lot of time that could be spent running "faster" tests). In this package, it is the only test and so it's fine to run it under nightly (skipping nightly stressrace because race builds with many nodes are very resource intensive and tend to collapse).